### PR TITLE
WIP Import user defined objects as modules

### PR DESCRIPTION
**Pull request by NDagestad**
This PR is on my packaging branch/PR.

This uses importlib's package metadata to make fuddly automagically discover data_models/projects and other user defined objects.

When a user defines a data_model{,_strategy} or project, they can add a [project.entry-points."<group-name>"] section in their pyporject.toml file. Fuddly can then request from importlib, all entry_points of a specific group-name.
2 group name were chose for fuddly:
- fuddly.data_models
- fuddly.projects

because of this, the requirement for project files to be named with a '_proj' suffix has been lifted.
If a project does have the suffix, it will be removed in the project name in fuddly's internal workings though.

The internal data_models/projects have been exposed in the fuddly's pyproject.toml file, and are therefore imported as modules instead of being looked up directly in the filesystem.

TODO

- [x] Documents the new behaviour
- [x] Merge data_models and data_models_strategies
        For now, a data_model, and data_model's strategy must have the same entrypoint name so fuddly can associate them. A better way of doing this would be if the strategies were added to an array of the data_model (similarly to how targets are defined for projects)
- [x] Check the filsystem before modules and warn but ignore duplicates
        This way, way can override a module by creating a local object with the same name.
- [x] Make fuddly detect if it is being run from a module or directly and import the internal object from the filesystem if ran directly
- [x] Put the entry point group names in a configuration or at least in a globally accessible variable

### CODE REVIEW

File: src/fuddly/framework/plumbing.py [hunk 1]
```diff
@@ -1071,6 +1274,7 @@ def _add_project(self, project, targets, logger, prj_rld_args, reload_prj=False)
             self.prj_list.remove(prj)
             self.prj_list.append(project)
         else:
+            # TODO: Catch this hieght up and ignore double imports ?
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py]: higher up ?

**Review by PoneyUHC — CHANGES_REQUESTED** [review]
typo

File: src/fuddly/framework/plumbing.py [hunk 2]
```diff
@@ -59,19 +59,18 @@
...
 
 import importlib
+from importlib.metadata import entry_points
 
 import io
 
 sys.path.insert(0, fuddly_data_folder)
 sys.path.insert(0, external_libs_folder)
 
+#TODO These are not needed right ?
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py]: to resolve

File: src/fuddly/framework/plumbing.py [hunk 3]
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
...
+            _, dirs, _ = next(os.walk(path))
+            for dirpath in dirs:
+                p = Path(os.path.join(path, dirpath))
+                inits = [ dirname(x) for x in  p.glob('**/__init__.py') ]
+                strats = [ dirname(x) for x in  p.glob('**/strategy.py') ]
+                modules = list(set(inits) & set(strats))
+                for m in modules:
+                    relpath = dirname(m)[len(base_path)+1:]
+                    key=prefix+relpath
+                    if data_models.get(key) is None:
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py:787]: use defaultdict ?

File: src/fuddly/framework/plumbing.py [hunk 4]
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
...
+            self.fmkDB.insert_dmaker(Database.DEFAULT_DM_NAME, Database.DEFAULT_GTYPE_NAME,
+                                     Database.DEFAULT_GEN_NAME, True, True)
+
+    def _get_data_models_from_fs(self, fmkDB_update=True):
+        if not self._quiet:
+            self.print(colorize(FontStyle.BOLD + "=" * 63 + "[ Data Models (filesystem) ]==", rgb=Color.FMKINFOGROUP))
+
         data_models = collections.OrderedDict()
 
         def populate_data_models(path, prefix=""):
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py:775]: why defining a function inside another to use it only once
--> get out or remove function

Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py:766]: not big fan of function in another, adds useless indentation (5-6 levels here)

Comment (NDagestad) [reply src/fuddly/framework/plumbing.py:766]: Not my addition, it was already like this before so I didn't change it

File: src/fuddly/framework/plumbing.py [hunk 5]
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
...
+            for name in names:
+                dm_params = self._import_dm(prefix, name)
+                if dm_params is not None:
+                    self._add_data_model(dm_params["dm"], dm_params["tactics"], dm_params["dm_rld_args"],
+                                         reload_dm=False)
+                    self.__dyngenerators_created[dm_params["dm"]] = False
+                    if fmkDB_update:
+                        # populate FMK DB
+                        self._fmkDB_insert_dm_and_dmakers(dm_params["dm"].name, dm_params["tactics"])
+                else:
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py]: if dm_params is None: false break

File: src/fuddly/framework/plumbing.py [hunk 6]
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
...
-                                     Database.DEFAULT_GEN_NAME, True, True)
+            if dm_params is not None:
+                self._add_data_model(dm_params['dm'], dm_params['tactics'], dm_params['dm_rld_args'], 
+                                     reload_dm=False)
+                name = dm_params['dm'].name
+                self.__dyngenerators_created[dm_params['dm']] = False
+                if fmkDB_update:
+                    # populate FMK DB
+                    self._fmkDB_insert_dm_and_dmakers(dm_params['dm'].name, dm_params['tactics'])
+            else:
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py]: same as before, if None: =false; break

File: src/fuddly/framework/plumbing.py (hunk 5 above)
Comment (NDagestad) [reply src/fuddly/framework/plumbing.py]: fmkDB_update is a boolean value?
dm_params is checked to not be None before entering the block?
I don't get what you are saying

File: src/fuddly/framework/plumbing.py (hunk 6 above)
Comment (NDagestad) [reply src/fuddly/framework/plumbing.py]: Same as before, fmkDB_update is a boolean

File: src/fuddly/framework/plumbing.py [hunk 7]
```diff
@@ -963,21 +1004,60 @@ def populate_projects(path, prefix=""):
...
+                                        f"ignoring... [{module.module}] ***", rgb=Color.WARNING))
+                continue
+
+            if prj_params is not None:
+                self._add_project(prj_params["project"], prj_params["target"],
+                                  prj_params["logger"], prj_params["prj_rld_args"],
+                                  reload_prj=False)
+                if fmkDB_update:
+                    self.fmkDB.insert_project(prj_params["project"].name)
+            else:
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py:1021]: same

File: src/fuddly/framework/plumbing.py [hunk 8]
```diff
@@ -963,21 +1004,60 @@ def populate_projects(path, prefix=""):
...
         except:
             if self._quiet:
                 return None
 
             if reload_prj:
-                self.print(colorize("*** Problem during reload of '%s_proj.py' ***" % (name), rgb=Color.ERROR))
+                self.print(colorize(f"*** Problem during reload of '{name}.py'/'{name}_proj.py' ***", rgb=Color.ERROR))
             else:
-                self.print(colorize("*** Problem during import of '%s_proj.py' ***" % (name), rgb=Color.ERROR))
+                self.print(colorize(f"*** Problem during import of '{name}.py'/'{name}_proj.py' ***", rgb=Color.ERROR))
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py:1060]: both branches are the same ??

File: src/fuddly/framework/plumbing.py (hunk 3 above)
Comment (NDagestad) [reply src/fuddly/framework/plumbing.py:787]: For what, data_model ? It is a `colloections.OrderedDict()` (based on what was used before)

File: src/fuddly/framework/plumbing.py [hunk 9]
```diff
@@ -1065,7 +1233,7 @@ def _add_project(self, project, targets, logger, prj_rld_args, reload_prj=False)
                 if prj.name == project.name:
                     break
             else:
-                raise ValueError
+                raise ValueError(porject.name)
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py]: c'est  toi le porject

Comment (NDagestad) [reply src/fuddly/framework/plumbing.py]: 🐷

File: src/fuddly/framework/plumbing.py [hunk 10]
```diff
@@ -1056,6 +1136,94 @@ def _import_project(self, prefix, name, reload_prj=False):
...
+            logger = module.logger
+        except AttributeError:
+            logger = Logger(name)
+        logger.fmkDB = self.fmkDB
+        logger.set_external_display(self.external_display)
+        if logger.name is None:
+            logger.name = name
+
+        # Targets
+        target=[]
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py]: wrong spelling, rip if error

Comment (NDagestad) [reply src/fuddly/framework/plumbing.py]: Python is magic, so it still works 🪄

File: src/fuddly/framework/plumbing.py [hunk 11]
```diff
@@ -1056,6 +1136,94 @@ def _import_project(self, prefix, name, reload_prj=False):
...
+        if not self._quiet:
+            if reload_prj:
+                self.print(colorize(f"*** Project '{name}' updated ***", 
+                                    rgb=Color.FMKSUBINFO))
+            else:
+                self.print(colorize(f"*** Found Project: '{name}' ***", rgb=Color.FMKSUBINFO))
+
+        return {
+            "project": project,
+            "target": targets,
```
Comment (PoneyUHC) [inline src/fuddly/framework/plumbing.py:1213]: "target" or "targets" ?

**Review by PoneyUHC — APPROVED** [review]

File: src/fuddly/framework/plumbing.py (hunk 11 above)
Comment (NDagestad) [reply src/fuddly/framework/plumbing.py:1213]: "target" = targets is correct

**Pull request merged**